server/src/main/java/org/apache/druid/discovery/DruidNodeDiscoveryProvider.java

```diff
@@ -66,10 +65,11 @@ public DruidNodeDiscovery getForService(String serviceName)
           if (nodeTypesToWatch == null) {
             throw new IAE("Unknown service [%s].", service);
           }
-
-          ServiceDruidNodeDiscovery serviceDiscovery = new ServiceDruidNodeDiscovery(service);
+          ServiceDruidNodeDiscovery serviceDiscovery = new ServiceDruidNodeDiscovery(service, nodeTypesToWatch.size());
+          DruidNodeDiscovery.Listener filteringGatheringUpstreamListener =
+              serviceDiscovery.filteringUpstreamListener();
           for (NodeType nodeType : nodeTypesToWatch) {
-            getForNodeType(nodeType).registerListener(serviceDiscovery.nodeTypeListener());
+            getForNodeType(nodeType).registerListener(filteringGatheringUpstreamListener);
           }
```

Comment:
It's mixes the concerns of `registerListener()`, making code harder to understand and more error-prone. It was part of the reason why I spent so much time understanding this code in the first place and why I created this PR.

Reply:
Now one should basically count all the listeners and set it in the constructor, looks more complex to me.

Reply:
Currently it's just `nodeTypesToWatch.size()` passed, hard to imagine how a mistake could be made. OTOH separation of concerns makes the whole code much more understandable.

Reply:
`Currently it's just nodeTypesToWatch.size() passed, hard to imagine how a mistake could be made.` Instead of one line someone should check two lines of code. Future on, those lines can be separated into different methods. 
Moreover, I think, if something can be done automatically it should be done that way.
And I don't see a good separation of concern here, it creates a possible situation when you add 2 listeners, but can specify in the constructor 1 as types count, does that makes sense at all?

Reply:
I don't understand about what one like you are talking here. I spent literally several hours wrapping my head around the former code.

You cannot just increment the `uninitializedNodeTypes` counter each time `registerListener()` is called, because this class it exposed to third-party users via `getForService()` method and they could add their own listeners. The initialization setup happens in `serviceDiscoveryMap.computeIfAbsent()` lambda and it set the fixed number of `uninitializedNodeTypes`.